pkg/controller/build/build_controller.go

```diff
@@ -239,67 +242,65 @@ func (r *ReconcileBuild) Reconcile(request reconcile.Request) (reconcile.Result,
 	if b.Spec.Output.SecretRef != nil && b.Spec.Output.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.Output.SecretRef.Name)
+		secretRefMap[b.Spec.Output.SecretRef.Name] = build.SpecOutputSecretRefNotFound
 	}
 	if b.Spec.Source.SecretRef != nil && b.Spec.Source.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.Source.SecretRef.Name)
+		secretRefMap[b.Spec.Source.SecretRef.Name] = build.SpecSourceSecretRefNotFound
 	}
 	if b.Spec.BuilderImage != nil && b.Spec.BuilderImage.SecretRef != nil && b.Spec.BuilderImage.SecretRef.Name != "" {
-		secretNames = append(secretNames, b.Spec.BuilderImage.SecretRef.Name)
+		secretRefMap[b.Spec.BuilderImage.SecretRef.Name] = build.SpecRuntimeSecretRefNotFound
 	}
 
-	if len(secretNames) > 0 {
-		if err := r.validateSecrets(ctx, secretNames, b.Namespace); err != nil {
-			b.Status.Reason = err.Error()
-			if updateErr := r.client.Status().Update(ctx, b); updateErr != nil {
-				// return an error in case of transient failure, and expect the next
-				// reconciliation to be able to update the Status of the object
-				return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
-			}
-			// The Secret Resource watcher will Reconcile again once the missing
-			// secret is created, therefore no need to return an error and enter on an infinite
-			// reconciliation
-			return reconcile.Result{}, nil
+	// Validate if the referenced secrets exist
+	if len(secretRefMap) > 0 {
+		if err := r.validateSecrets(ctx, secretRefMap, b); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if validationFailed {
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 	}
 
-	// Validate if the build strategy is defined
+	// Validate if the referenced strategy exists
 	if b.Spec.StrategyRef != nil {
-		if err := r.validateStrategyRef(ctx, b.Spec.StrategyRef, b.Namespace); err != nil {
-			b.Status.Reason = err.Error()
-			updateErr := r.client.Status().Update(ctx, b)
-			return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
+		if err := r.validateStrategyRef(ctx, b); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		if validationFailed {
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 		ctxlog.Info(ctx, "buildStrategy found", namespace, b.Namespace, name, b.Name, "strategy", b.Spec.StrategyRef.Name)
 	}
 
-	// validate if "spec.runtime" attributes are valid
+	// Validate "spec.runtime" attributes
 	if utils.IsRuntimeDefined(b) {
-		if err := r.validateRuntime(b.Spec.Runtime); err != nil {
-			ctxlog.Error(ctx, err, "failed validating runtime attributes", "Build", b.Name)
-			b.Status.Reason = err.Error()
-			updateErr := r.client.Status().Update(ctx, b)
-			return reconcile.Result{}, fmt.Errorf("errors: %v %v", err, updateErr)
+		if r.validateRuntimeFailed(b) {
+			return r.UpdateBuildStatusAndRetreat(ctx, b)
 		}
 	}
 
 	b.Status.Registered = corev1.ConditionTrue
+	b.Status.Message = build.AllValidationsSucceeded
```

Comment:
Do we also need to set the Reason to `nil` to remove the error code?

Reply:
@SaschaSchwarze0 
If all validations are successful, then `b.Status.Reason = build.SucceedStatus`. This is defined at the beginning of reconcile: https://github.com/qu1queee/build/blob/refactor_build_fields/pkg/controller/build/build_controller.go#L245
So I think we don't need to set the Reason to be `nil`.

Reply:
I see, thank you.